### PR TITLE
Added new modules and es6 module system

### PR DESCRIPTION
**Pull request by w0298i**
Opening a new PR, as the last one is no longer relevant. I've nearly finished refactoring MochiKit, but I haven't rebuilt some modules correctly. I would appreciate any help, as this is a big project.

### CODE REVIEW

**Comment by w0298i** [comment]
update: built mochikit, working atm

**Comment by w0298i** [comment]
ive been working on this all day, need some help with the test system @etrepum

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+<<<<<<< HEAD
+    "python.linting.enabled": false
+=======
+    "todo-tree.flat": false,
+    "todo-tree.expanded": true,
+    "npm-scripts.showStartNotification": false
+>>>>>>> 82b1904cd94eab193a95b089863079fbebbe174e
+}
```
Comment (etrepum) [inline .vscode/settings.json]: This file appears to have been committed with an unresolved conflict

File: MochiKit/MochiKit.js [hunk 2]
```diff
@@ -7,127 +7,34 @@ See <http://mochikit.com/> for documentation, downloads, license, etc.
...
-    "Position",
-    "Visual",
-    "DragAndDrop",
-    "Sortable"
-];
+if (typeof console.log === 'function') {
+    let {
+        version: { major, minor, patch }
+    } = MochiKit;
+    console.log(`MochiKit version ${major}.${minor}.${patch} loaded!`);
```
Comment (etrepum) [inline MochiKit/MochiKit.js]: It's usually discouraged to have a log like this in a production library

File: .vscode/settings.json (hunk 1 above)
Comment (w0298i) [reply .vscode/settings.json]: I'll add it to the gitignore tmrw, thx

**Comment by blq** [comment]
Interesting new project. I'll try to follow and give some feedback.

File: MochiKit/async/Deferred.js [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    return this;
+};
+
+Deferred.prototype.tap = function(func) {
+    return this.then(function(value) {
+        //Prevent the value from being changed
+        //by the callback, as it is a lazy function
+        //meaning it performs operations without
+        //returning a value.
+        func(value);
```
Comment (blq) [inline MochiKit/async/Deferred.js]: A slightly different semantic for 'tap' is that if 'func' returns a Promise the chain waits for it. Bluebird does that for example.

Comment (w0298i) [reply MochiKit/async/Deferred.js]: Interesting. This was heavily inspired by Bluebird,

**Comment by w0298i** [comment]
@blq You still interested in giving feedback? The fork is at w0298i/mochikit.

**Comment by w0298i** [comment]
Th best way to *help* me is to give me feedback. Judging by Bob's activity log, I can understand he isn't very interested in MochiKit which is fine, as a lot of time has passed (13 years). I'd just like to know exactly where I'm at with you guys.

**Comment by etrepum** [comment]
I’d really like to see tests and have them run in CI (Travis is probably the easiest to set up). What you’ve done so far looks like it has a lot of potential but there’s no easy way to know if any of it works. Code review isn’t really a good enough QA process for a change set this large when it’s done by hand.

**Comment by w0298i** [comment]
It can be built using Rollup, and tested with Mocha. I've used mocha in the past for my log.ts project, and I can confirm it's great. It uses `describe(...) it(..., test)` syntax.

**Comment by w0298i** [comment]
I'm also planning to add the methods at @blq 's fork to the main repo.

**Comment by w0298i** [comment]
I don't know Travis though, I'm interested in learning it if you think it's worth it.

**Comment by etrepum** [comment]
Mocha is fine, having tests is far more important than which testing libraries are used.

I very strongly recommend learning how to set up Travis CI. It’s relatively painless and saves a whole lot of manual effort once it’s there.

**Comment by w0298i** [comment]
I've added Travis tests. I've learned the basics, which are very minimal. Glad I don't have to learn another lang lol. Now to add the *.test.js files...

**Comment by blq** [comment]
@w0298i  yeah, I'll try to read up. Nice that you found my fork, was going to suggest looking at some parts of it :) 
Looks like you done much with Async lib. Do you maybe have a more high-level plan for that one? Did you mean it will still be backwards-compatible (apart from some deleted methods) or some hybrid?

**Comment by w0298i** [comment]
@blq Probably a hybrid. I did try to keep it close to the old async, but I've probably missed a few things. I think the hardest think to revamp was the Base, but the Visual class looks very large. I also added an `all.js` file, which doesn't name the categories. Everything is exported in one namespace. You can also check out my `todo` file in `/`, which is all the things I'm gonna do :) You should also check out my [gh pages site](https://w0298i.github.io), which suggests I'm not too bad at making barebones sites in es6, html and css. If you want to help me, you could give feedback or make a fork of my fork, then submit a pull request. I have optimised MochiKit too, killing off unnecessary `typeof` checks for `undefined || null`, replacing it with `arg == null`. That `datetime` class is looking a bit lonely, too. Might need to add some more time-based methods onto it, as there isn't a lot on there right now. I also found some links to ` http://svn.mochikit.com/mochikit/trunk`, not sure what that is completely, but you might want to know that `mochikit.com` is dead, so you should link to the gh page instead. Think thats it lol :+1:

**Comment by w0298i** [comment]
@etrepum Does the `Signal` module have inspiration from another module, maybe a `Python` module? Just wondering what the rules are for making it.

**Comment by w0298i** [comment]
So.. another question. Is MochiKit just a port of the best python modules? Looks a bit like it to me tbch.

**Comment by etrepum** [comment]
MochiKit predates GitHub and Google Code, so there may be some references to a very old self-hosted Subversion repository.

I’m not the primary author of Signal but the answer to your question is right in the docs:
> This event system is largely based on Qt's signal/slot system

Where possible, MochiKit took inspiration from Python’s standard library and libraries we used such as Twisted. However, very little of it is an actual port and some of it comes from elsewhere (Qt, AppKit, Perl, etc.) or was completely original based on our needs.

**Comment by w0298i** [comment]
Cool. @etrepum Like the build scripts?

**Comment by w0298i** [comment]
I'm actually planning to add more to DOM too, as it's looking a bit slim.

**Comment by w0298i** [comment]
MochiKit looks better than ever here. I'm excited to see what the future holds.

**Comment by etrepum** [comment]
Not sure what you mean "like the build scripts". Those are all bespoke, and written in Python because at the time node didn't even exist and it was extremely uncommon to run JavaScript from the command line.

**Comment by w0298i** [comment]
@etrepum I meant "do you like them" because i re-wrote them lol.

**Comment by etrepum** [comment]
The rollup based build looks nice. Doesn't appear to be compatible with any existing MochiKit code due to the renaming (MochiKit -> mochikit) but that is probably not an issue at this point.

**Comment by w0298i** [comment]
Yeah, I followed d3 with the category builds. Great for people that only want the "async" class for example, without the overhead of other modules. Viewed the screencast lastnight, seems like MochiKit was great for it's time. We should probably make a <video> element for it though, so more people watch it. I'll submit a PR to yours tonight fixing that.

**Comment by w0298i** [comment]
So... I was thinking of keeping the older docs, for retro coolness. What do you think @etrepum?

**Comment by w0298i** [comment]
gotcha gotcha - can you research the best tools for the job?

**Comment by w0298i** [comment]
gah sorry about the useless fork - schools been getting on top of me. ill add it to my fork today

**Comment by blq** [comment]
Looked more at your fork and

* What I always liked about MK and that I think is big part of its "identity" is the Python style. I think we should continue with that path, i.e when in doubt choose the Python-names, API, philosophy. It gives us really good references for docs and best practices. (altough I'm not a Python expert! I also like C++/STL/Boost stuff that I think are good inspiration for APIs and very underused in JS world)

* Before adding too much I think we should see what to remove, distill the best parts so that MK will have an edge on those, not do it all in similar way as so many other libs already do. So not just "fatten up" modules for the sake of it ;)  

* I think you are doing mistakes in the iter.js code. It should focus on lazy-eval and here we should migrate to use ES6 Iter standard (unfortunately away from the previous Python-iter-throw-convention)
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
I'd say the high-level functions would be possible to port without breaking things(?). Then maybe some converter functions for "deep" conversions (I guess interacting directly with the StopIter is quite rare?)
(In my fork I added stuff from Python's itertools and C++ STL etc http://blq.github.io/mochikit/doc/html/MochiKit/Iter-ext.html )

* based on "old" MK a hard prune I'd say could be something like this?

- Async => keep (move to ES6 Promise back-end)
- Base => keep (except stuff in ES5+)
- DOM => drop 50%, coerceToDOM and createDOM is the core
- DragAndDrop => drop
- Color => drop 
- DateTime => drop
- Format => drop
- Iter => keep (move to ES6 Iter back-end)
- Logging => drop
- LoggingPane => drop
- Selector => drop
- Signal => keep (could we improve GC for this?)
- Style => drop
- Sortable => drop
- Text => keep(?)
- Visual => drop

More ideas coming :)

(Do we maybe have some better place to discuss and take notes? (wiki?) )

**Comment by w0298i** [comment]
I was trying to for a more oop style iteration technique, I'll look into what you said with greater detail. I'm not sure I completely agree with you, though. We could keep the Logging & Sortable class. I tried to go for a more generator-style iteration technique, I haven't used iteration like too much though, so I'm not sure if `Generator#next` is a factory funvtion that returns an object, or if the original generator is modified with the `#next` value and `done` prop. Thanks for the feedback though. If you have any ideas floating about, be sure to have a fork of my fork (lol) on hand, so you can submit PRs to me with any extra functionality. Thanks again 🐈 < my cat :)

**Comment by blq** [comment]
Read more here for example https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#The_iterator_protocol
It's not a good idea to invent another protocol when we now even have ES6 standard. And as said, operations in such a module would be assumed to be fully lazy, i.e no up-front arrayification or such.
And as said, if an operation already exists in Pyton we should basically copy that name and functionality IMHO.

Not super strong opinion on Sortable, just that I'd suspect a focused lib for that exists and might not be an edge for MK.

A wiki might be good to add these kind of notes and "project plans" rather than in this more temporary thread.

**Comment by blq** [comment]
In my https://github.com/blq/mochikit/blob/master/MochiKit/Iter-ext.js  I have a toES6Iterator() and fromES6Iterator by the way where you can see the relatively simple change of pattern for the protocol.

**Comment by w0298i** [comment]
So youre saying to completely ditch the MochiKit iter protocol?

**Comment by blq** [comment]
Kindof, I mean we should migrate to ES6 protocol but try to stay backwards compatible apart from lowest-level StopIter stuff. I think it's possible.

**Comment by w0298i** [comment]
es6 protocol... ok... what about an `IteratorValue` class for yeilded values?

**Comment by blq** [comment]
Not sure I understand. ES6 you return { value: value, done: boolean } instead of in MochiKit/Python return value+throw-StopIteration-if-done   (generators and "yield" keyword is not necessary for this)

**Comment by w0298i** [comment]
I mean instead of returning an object, maybe an IteratorValue would be shorter to type.

**Comment by w0298i** [comment]
And it's a pain to copy the __repr__ stuff across plain objects, a unary __repr__ value would be best.

**Comment by blq** [comment]
I mean the return from the next() method.
(I could image dropping repr support actually. browser debuggers are much better today, useful, but less need for it)

**Comment by etrepum** [comment]
I agree that Deferred should be replaced entirely with Promise, but that there’s plenty of room for helpful functions in the async module.

The MochiKit iteration protocol should be removed in favor of the ES6 standard and all functions should be refactored accordingly.

Logging could go or at least get a lot simpler as the console object is basically standardized.

Most of Visual can probably go as that stuff can nearly all be done in CSS now. The stuff that can’t should probably have a more modern rAF approach.

There’s got to be a better way to do Signal by now. After writing a little bit of code with it recently I would prefer something a little closer to the native Event object like React does.

I don’t see a reason to drop Color, none of that functionality is natively available.

Don’t have time right now to comment on other modules.

**Comment by w0298i** [comment]
Why drop Deferred? Have you tried using it yet? It has a lot more utility functions than the native `Promise`.

**Comment by etrepum** [comment]
I’m quite familiar with both Deferred and Promise. Deferred should no longer exist, but the useful parts of it can be made available as functions. Standardization and syntax support trumps any convenience that the additional methods provide.

**Comment by w0298i** [comment]
Let's not be too eager to drop things, as we need to keep what made MochiKit MochiKit in my opinion (so far). We need to stretch the examples too, as more code should - in user's eyes - have more functionality.

If you have any modules to suggest, then throw them at me. I'll find out what DOM/JS devs hate and build on that. That's a good place to start.

In reply to "fattening up the DOM class" I do think that's a valid issue, and not just me saying let's add functionality for the hell of it. If a class has little functionality, then it isn't enough in my eyes.

As I said, I'll stop blindly reconstructing classes in es6, and see what is needed right now. Seeing as MochiKit is 14ish years old, It's a good mindset to have.

In reply to your comment just now, I'll consider methodizing the functions, although some of them use Deferred-only functionality.

**Comment by w0298i** [comment]
We could even make a class that extends a native Promise and rely more on native functionality...

**Comment by etrepum** [comment]
There’s nothing that Deferred does that can not be done with Promise. We should be using functions, not classes and methods, to deal with most things (particularly related to async and iteration).

**Comment by w0298i** [comment]
Oh by methods I mean functions btw

**Comment by w0298i** [comment]
Maybe we could add SVG drawing support with algorithms?

**Comment by w0298i** [comment]
Are you interested in making a DOM chainable class API like jQuerys @blq @etrepum? That might be a strong base to go off - as jQuery is a very bloated library, maybe we could offer a cleaner alternative?

**Comment by etrepum** [comment]
I never liked the chainable style, or having to wrap every Element/NodeList you want to work with. My preference would be to have other ways to compose DOM traversal and manipulation, working a bit more natively. The thing that jQuery really got right in my opinion was working with selectors.

**Comment by w0298i** [comment]
You got one thing *completely* wrong in your previous `Selector` module. You used `eval`. I agree though that chaining is a bit silly. I have made lots of jQuery clones that have more fine grained support like `setAttribute` instead of just `attr`. It's a bit *too* simple, in that respect.

**Comment by w0298i** [comment]
I ain't doing bad in this whole thing to tbh lol, a lot of time and love has gone into this rofl

**Comment by w0298i** [comment]
Should I port over the best python modules? I got some on my phone I thought we could throw in:
```
python calendar
python warning
node stream
node readline (maybe not)
java scanner
```

Those are some ideas. We should really ship the best modules out of each language.

**Comment by etrepum** [comment]
At the time using eval worked better than not (for performance reasons). JS engines have improved a lot since then. It's not necessary at all anymore given querySelectorAll.

**Comment by w0298i** [comment]
Yes. Any modules you like using from other langs @etrepum (that others do too)? Would be happy to port them.

**Comment by etrepum** [comment]
I think it makes sense to start with a smaller core and make sure that it's well tested and documented, and then add more as necessary. Generally speaking all of the modules in MochiKit are there because we needed them to build the apps we were building. Typically the workflow was to build an app-specific version of some functionality and then once it proved useful in more than one place we'd extract it and add it to MochiKit.

Outside of a commercial setting it would probably make sense to start from an example that you'd like to build, and build it without adding any code to MochiKit itself. Once it works, you can take a look to see if anything in there is generally applicable to other uses and consider extracting it from the example and adding it to MochiKit.

**Comment by w0298i** [comment]
gah ok ill do the tests lol, ill start on them now

**Comment by etrepum** [comment]
The tests are the most important part :)

**Comment by blq** [comment]
I see you refactored iterators now, good. But I think this IteratorValue class is not a good idea. Introducing a full class, that also includes an optional back-reference and a repr() method is unnecessarily heavy. Simply returning a literal with done and value is fine.
Remember that the iteration-functions are low-level functions not really intended to be (re)written by the end-user programmer, later you use the high-level functions. As MK Iter says in its one-liner: " iteration made HARD, and then easy" ;)

**Comment by blq** [comment]
@etrepum yeah, I pruned hard, much based on my own usage. Color is of course not affected by browser-standards and new tech and could be kept.

But, I'd like to hear more high-level visions here. 
- Should we strive for backwards compatibility? 
I think we could add new Promise-stuff both "inside" Deferred and parallel to it. For example; simply making Deferred "thenable" would make it basically seamlessly usable in async/await code without affecting old code. 
Similarly as mentioned, migrating Iteration to ES6 protocol would still keep almost all old MK high-level iter code functioning. 

My suggestion for Signal is to drop the DOM-stuff, there some natvive or react-ish style code might be best fit today. But for signalling between app-components in a program I'd say the Signal API is really good. Wish-list is better performance for large number of signals and "something" to help cleanup/dispose of them (why don't JS have destructors..!!)

**Comment by blq** [comment]
Speaking of Promise utilities with only functions, I have an experimental stash here: https://github.com/blq/blq/blob/master/promise.js
Inspired by Bluebird etc, also includes bridges to MK and jQuery etc.

**Comment by w0298i** [comment]
I think it's a good idea. I'll think about it.

**Comment by w0298i** [comment]
Have you even seen my Deferred class? Much better than a native Promise. I don't know. Would like a 2nd opinion.

**Comment by w0298i** [comment]
How is it heavy lol, it's just a class.

**Comment by w0298i** [comment]
Could we maybe make a DOM adapter for the Signal class? I've found it better not to bake things like that in to the core. Maybe we could add a `dispose` method to all classes that hold data too. I'll keep working hard on these tests then.

**Comment by etrepum** [comment]
Using a class is just not very composable, you’ll end up having to wrap and unwrap things a lot since most things are going to be using a native Promise.

**Comment by w0298i** [comment]
@etrepum What do you mean? Are you talking about my `IteratorValue` class or `Deferred`?

**Comment by blq** [comment]
that "just a class" (IteratorValue) will probably be the one that is created and destructed, taxing the GC, most of all objects in your program, every iteration step, millions of times. Sure actual profiling tells the thruth, but I'd say this is a good example were staying lean form the start is best, the win with that class is marginal also.

**Comment by blq** [comment]
@etrepum I would be interested to learn about some examples of your async ideas. Links? 
In my little lib https://github.com/blq/blq/blob/master/promise.js I created most base functionality as pure functions for Promises, but I can image you had something more in mind anyway?  (I agree that chainable apis are often overused, but async-style then-then-then.. is a tricky place to enter(?))

**Comment by w0298i** [comment]
If you want to stop using the IteratorValue class, feel free to submit a PR on my fork. I'm too busy with the tests, so outside help would be good if you want that done. @blq

**Comment by blq** [comment]
Sure, but a bit difficult to grasp the state of your code, your're moving fast ;)
TBH I think this PR should be split in several smaller changes instead so we can have a more focused evaluation.

**Comment by w0298i** [comment]
Rome wasn't built in a day, but this code might be lol. 

PR-wise, just throw one at me that removes this whole `IteratorValue` thing if you can.

**Comment by w0298i** [comment]
Just let me push these changes...

**Comment by w0298i** [comment]
PR-away...

**Comment by w0298i** [comment]
On a side note, this is the first successful travis build...

**Comment by w0298i** [comment]
Could we have some sort of speed test fir the modules in node? Just an idea. Maybe like a profiler tool. We might have to build one, though.

**Comment by etrepum** [comment]
I’d start with tests, coverage reporting, and docs… then worry about benchmarking.

**Comment by w0298i** [comment]
It's purely an idea for later. @blq How's that PR going?

**Comment by w0298i** [comment]
It doesn't seem to have come through.

**Comment by w0298i** [comment]
@etrepum Should we get some kind of internal chat? This is getting slower and slower to load, and it feels a bit messy.

**Comment by blq** [comment]
I looked more at your iter. code and that specific class is not really the biggest issue now. 
My take on iter is that is should keep the existing MK API (including the tests). Your code changes "everything" here without sufficient motivation IMHO. We need to take a step back and re-evaluate, or postpone it from this PR. 
I'm currently looking into ways to keep compatibility with ES6 iters. I could put that in my fork for now.

**Comment by w0298i** [comment]
I ditched the MochiKit API for the ES6 Iteratror spec. It's time to move on. This was written *10* ywars ago, remember.

**Comment by w0298i** [comment]
I ported all of the functions to es6. If you want to add a `toES6Iterator` pipe for legacy `Iter` methods, be my guest.

**Comment by w0298i** [comment]
You could even port your functions, from your `blq/blq` repo.

**Comment by etrepum** [comment]
Here’s a discord server we can use https://discord.gg/EBbKAfv

**Comment by w0298i** [comment]
Shit, [Discord](https://www.reddit.com/r/discordapp/comments/7a7vq3/discord_privacy_policy_shows_users_data/). Would you be interested in using Riot?

**Comment by blq** [comment]
The ES6 Iterator spec is simply about how to take the "next step" at lowest level. Nothing about higher-level functions. It's possible to migrate and keep existing API. Please motivate your new function APIs. Existing practice, simple links to some other language or libs you got the ideas from?
The old API lives on fine in Python 3 more or less.
https://docs.python.org/3/howto/functional.html
If you drop all of that it's not MK anymore IMO.

**Comment by w0298i** [comment]
If you want the old features, make a fork. I'm moving on.

**Comment by blq** [comment]
Sigh, those functions/API are Not old. 
Here's some "modern" libs today for example:
https://lodash.com/docs/4.17.10
https://underscorejs.org/
Concepts are same since 60s functional programming. No need to make up new names for same-ish stuff.

**Comment by w0298i** [comment]
I'll do my bit, and anything that's wrong you could send in a PR for.

**Comment by w0298i** [comment]
I've made more lazy functions in `Iter`. Check them out! :) @blq @etrepum

**Comment by blq** [comment]
hmm, which ones? cycle?
(your cycle changes behaviour from typical implementation that copies input, it only works for Array)

**Comment by w0298i** [comment]
It also works with array-like objects. Maybe add support for Maps/Sets?

**Comment by etrepum** [comment]
On my phone so I can’t really do a thorough review but it looks like cycle and/or chain has the wrong semantics (without even looking at the base class, these are different functions and shouldn’t have the same implementation). I’d take a look at the 1.x tests for all of these functions and make sure they work the same way (modulo using ES6 iterators instead of MochiKit’s iter protocol). You can also refer to the Python itertools module and its tests.

**Comment by w0298i** [comment]
I'm going to take a break from MochiKit. @blq You thinking of adding those ideas you mentioned to my fork with a PR?

**Comment by w0298i** [comment]
I've fucked something up **big time** online. Good luck with MochiKit, guys.

**Comment by def14nt** [comment]
Nice! The problem i've encountered and probably solved is storing data when im using a file for each function. I'll put it back like it was, with the old layout. Tree shaking will still work, but it'll be less chaotic to search the directories.

**Comment by etrepum** [comment]
I think the best approach to reorg would be to take the existing code and apply a series of transformations to it (a la jscodeshift and prettier or similar). Would provide a good baseline to do by-hand improvements in a way that’s easier to review.

**Comment by def14nt** [comment]
tf is jscodeshift? sounds interesting... ill have a look

**Comment by def14nt** [comment]
wait, why transformations? doesnt make sense to me tbch looking at their docs

**Comment by etrepum** [comment]
I do like the approach of using WeakMap for signal, probably won’t have a chance to take a close look at the code for a few weeks due to travel

**Comment by etrepum** [comment]
A large amount of the modernization we want to do to MochiKit can be done mechanically. It’s much easier to review a transformation than to review thousands of individual changes (any of which could have a typo).

**Comment by def14nt** [comment]
interesting. @blq Something for you to think about: I have like two *other* projects to maintain, so could you lend a hand with the code - being transformations - here?

**Comment by def14nt** [comment]
@etrepum Could you contribute anything?

**Comment by def14nt** [comment]
No worries. I'll see what I can do.

**Comment by def14nt** [comment]
# Sorry

I'm going to leave a link to the [code I wrote](https://mega.nz/#!l65hGBib!1IJ2o4r-RcJ1Bi1XTa9ft33oyFBGEI7cIRpmT1DApz4) in this chat. I've had a think about it, and I'm not going to contribute to something that doesn't have *any* community pull, and I just don't want to work on it for reasons which may be obvious to some who I have communicated with on other platforms. I've left a zip file in here, so you can still write code for MochiKit. I wish you all the best with your library, and whatever else you do. You may seem me as petty, but with my education I'm not keeping up with things, and the time I have I should have been spending on thing I enjoyed to begin with.

From my eyes, MochiKit was built for an older time, where the JS engines were sub-par, so a refactor might be a complete waste of time to begin with. I don't think this is a strong idea at all, and I don't see MochiKit fitting in anymore.

It may be obvious that I will *not* have anything else to do with this library from this point in time, and I will be revoking my likes/watches from all libraries authored by all contributors of this library. **The link will not be altered in any way.**. I'm also very excited to see @etrepum and @blq 's ideas get added into upstream. Good luck with that, too.

**Comment by def14nt** [comment]
@blq Sorry you feel that way. If you want to add more to the new version (I doubt it) then you can.

**Comment by def0x099** [comment]
Once again, I'm **really** sorry for my behaviour here. It was appalling and it was very uncalled for. In retrospect, I wasn't thinking straight (which was obvious at the time) and I really am shocked at some of the shit I've come out with. 

I agree with @blq. My attitude was unrealistic looking back at this, but that may be due to lack of education :p
Anyway, thanks for reading this. Hope you all had a great Christmas! :santa: :mrs_claus:

**Pull request closed without merging**